rust/src/initramfs.rs

```diff
@@ -0,0 +1,117 @@
+//! Generate an "overlay" initramfs image
+
+use anyhow::{Context, Result};
+use openat::SimpleType;
+use std::collections::BTreeSet;
+use std::collections::{HashMap, HashSet};
+use std::fs;
+use std::io;
+use std::io::prelude::*;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::process::CommandExt;
+
+fn add_etc_files_recurse(
+    d: &openat::Dir,
+    path: &str,
+    filelist: &mut BTreeSet<String>,
+) -> Result<()> {
+    let meta = d.metadata(path).context("stat")?;
+    match meta.simple_type() {
+        SimpleType::Symlink | SimpleType::File => {}
+        SimpleType::Dir => {
+            for e in d.list_dir(path).context("readdir")? {
+                let e = e.context("readdir")?;
+                let name = e.file_name();
+                let name = match name.to_str() {
+                    Some(n) => n,
+                    None => anyhow::bail!("Invalid UTF-8 name {}", name.to_string_lossy()),
+                };
+                let subpath = format!("{}/{}", path, name);
+                add_etc_files_recurse(&d, &subpath, filelist)?;
+            }
+        }
+        _ => anyhow::bail!("Invalid non-regfile/symlink/directory: {}", path),
+    }
+    filelist.insert(path.to_string());
+    Ok(())
+}
+
+fn generate_initramfs_overlay(files: &HashSet<String>) -> Result<fs::File> {
+    let etcd = openat::Dir::open("/etc")?;
+    let mut filelist: BTreeSet<String> = BTreeSet::new();
+    for file in files {
+        let file = match file.strip_prefix("/etc/") {
+            Some(f) => f,
+            None => anyhow::bail!("Invalid non-/etc path: {}", file),
+        };
+        add_etc_files_recurse(&etcd, file, &mut filelist)
+            .with_context(|| format!("Adding {}", file))?;
+    }
+    let mut out_tmpf = tempfile::tempfile()?;
+    {
+        let mut proc = std::process::Command::new("/bin/sh");
+        proc.arg("-c")
+            .arg("cpio --create --format newc --quiet --reproducible --null | gzip -1")
```

Comment:
I am looking at the https://docs.rs/subprocess/0.2.6/subprocess/ crate which looks like it has all the interfaces we need to make this better and avoid the unsafe nix bits below.